src/Autofac.Pooling/RegistrationExtensions.cs

```diff
@@ -0,0 +1,369 @@
+        /// <param name="lifetimeScopeTags">Tags applied to matching lifetime scopes.</param>
+        /// <returns>The registration builder.</returns>
+        public static IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle>
+               PooledInstancePerMatchingLifetimeScope<TLimit, TActivatorData, TSingleRegistrationStyle>(
+                   this IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle> registration,
+                   IPooledRegistrationPolicy<TLimit> poolPolicy,
+                   params object[] lifetimeScopeTags)
+               where TSingleRegistrationStyle : SingleRegistrationStyle
+               where TActivatorData : IConcreteActivatorData
+               where TLimit : class
+        {
+            if (registration == null) throw new ArgumentNullException(nameof(registration));
+
+            RegisterPooled(registration, poolPolicy, lifetimeScopeTags);
+
+            return registration;
+        }
+
+        private static void RegisterPooled<TLimit, TActivatorData, TSingleRegistrationStyle>(
+            IRegistrationBuilder<TLimit, TActivatorData, TSingleRegistrationStyle> registration,
+            IPooledRegistrationPolicy<TLimit> registrationPolicy,
+            object[]? tags)
+            where TSingleRegistrationStyle : SingleRegistrationStyle
+            where TActivatorData : IConcreteActivatorData
+            where TLimit : class
+        {
+            if (registration == null) throw new ArgumentNullException(nameof(registration));
+
+            // Mark the lifetime appropriately.
+            var regData = registration.RegistrationData;
+
+            regData.Lifetime = new PooledLifetime();
+            regData.Sharing = InstanceSharing.None;
+
+            var callback = regData.DeferredCallback;
+
+            if (callback is null)
+            {
+                throw new NotSupportedException(RegistrationExtensionsResources.RequiresCallbackContainer);
+            }
+
+            if (registration.ActivatorData.Activator is ProvidedInstanceActivator)
+            {
+                // Can't use provided instance activators with pooling (because it would try to repeatedly activate).
+                throw new NotSupportedException(RegistrationExtensionsResources.CannotUseProvidedInstances);
+            }
+
+            var original = callback.Callback;
+
+            Action<IComponentRegistryBuilder> newCallback = registry =>
+            {
+                // Only do the additional registrations if we are still using a PooledLifetime.
+                if (regData.Lifetime is PooledLifetime)
+                {
+                    var pooledInstanceService = new UniqueService();
+
+                    var instanceActivator = registration.ActivatorData.Activator;
+
+                    if (registration.ResolvePipeline.Middleware.Any(c => c.ToString() == nameof(Autofac.RegistrationExtensions.OnRelease)))
```

Comment:
It seems scary to me to compare middleware by string representation, like it's not going to be refactor-friendly. Is there another way?

Reply:
Fair point.

I think we may have to change Autofac v6 to have distinct types for delegate middleware that is considered 'core'.

I'll look at raising a PR over there.

Reply:
OK, let's leave this as it is for now and we can fix it post-update in the other repo.